Plugins/GRPCProtobufGeneratorCommand/Plugin.swift

```diff
@@ -0,0 +1,245 @@
+        protocGenSwiftPath: protocGenSwiftPath,
+        outputDirectory: outputDirectory
+      )
+
+      let completionStatus = try executeProtocInvocation(
+        executableURL: protocPath,
+        arguments: arguments,
+        verbose: commandConfig.verbose,
+        dryRun: commandConfig.dryRun
+      )
+
+      if !commandConfig.dryRun {
+        Stderr.print(
+          "Generated protobuf message Swift files for \(inputFiles.joined(separator: ", "))."
+        )
+      }
+    }
+  }
+}
+
+/// Execute a single invocation of `protoc`, printing output and if in verbose mode the invocation
+/// - Parameters:
+///   - executableURL: The path to the `protoc` executable.
+///   - arguments: The arguments to be passed to `protoc`.
+///   - verbose: Whether or not to print verbose output
+///   - dryRun: If this invocation is a dry-run, i.e. will not actually be executed
+func executeProtocInvocation(
+  executableURL: URL,
+  arguments: [String],
+  verbose: Bool,
+  dryRun: Bool
+) throws {
+  if verbose {
+    Stderr.print("\(executableURL.absoluteStringNoScheme) \\")
+    Stderr.print("  \(arguments.joined(separator: " \\\n  "))")
+  }
+
+  if dryRun {
+    return
+  }
+
+  let process = Process()
+  process.executableURL = executableURL
+  process.arguments = arguments
+
+  let outputPipe = Pipe()
+  let errorPipe = Pipe()
+  process.standardOutput = outputPipe
+  process.standardError = errorPipe
+
+  do {
+    try process.run()
+  } catch {
+    try printProtocOutput(outputPipe, verbose: verbose)
+    let stdErr: String?
+    if let errorData = try errorPipe.fileHandleForReading.readToEnd() {
+      stdErr = String(decoding: errorData, as: UTF8.self)
+    } else {
+      stdErr = nil
+    }
+    throw CommandPluginError.generationFailure(
+      errorDescription: "\(error)",
+      executable: executableURL.absoluteStringNoScheme,
+      arguments: arguments,
+      stdErr: stdErr
+    )
+  }
+  process.waitUntilExit()
+
+  try printProtocOutput(outputPipe, verbose: verbose)
+
+  guard process.terminationReason == .exit && process.terminationStatus == 0 else {
+    let stdErr: String?
+    if let errorData = try errorPipe.fileHandleForReading.readToEnd() {
+      stdErr = String(decoding: errorData, as: UTF8.self)
+    } else {
+      stdErr = nil
+    }
+    let problem = "\(process.terminationReason):\(process.terminationStatus)"
+    throw CommandPluginError.generationFailure(
+      errorDescription: problem,
+      executable: executableURL.absoluteStringNoScheme,
+      arguments: arguments,
+      stdErr: stdErr
+    )
+  }
+
+  return
+}
+
+func printProtocOutput(_ stdOut: Pipe, verbose: Bool) throws {
+  let prefix = "\t"
+
+  if verbose, let outputData = try stdOut.fileHandleForReading.readToEnd() {
+    let output = String(decoding: outputData, as: UTF8.self)
+    let lines = output.split { $0.isNewline }
+    print("protoc output:")
+    for line in lines {
+      print("\(prefix)\(line)")
```

Comment:
There's no point pulling `prefix` out, it's just unnecessary indirection here

```suggestion
  if verbose, let outputData = try stdOut.fileHandleForReading.readToEnd() {
    let output = String(decoding: outputData, as: UTF8.self)
    let lines = output.split { $0.isNewline }
    print("protoc output:")
    for line in lines {
      print("\t\(line)")
```